### DISCUSSION TITLE
How to avoid running Source Generator using CompilationProvider on every keystroke

### DISCUSSION THREAD

**Discussion by Dreamescaper** [dotnet/roslyn#74001] [Q&A]
Hi all,

I have created a [Source Generator](https://github.com/Dreamescaper/ServiceScan.SourceGenerator/) to assist with DI services registration. The idea is that it scans assembly for types based on a criteria in an user-defined attribute.

So currently the pipeline is the following:
1. Get the cacheable model from the attribute with search criterias.
2. Combine with CompilationProvider.
3. Get the assembly from Compilation, scan types from this assembly, get cacheable model of the method to generate based on found types.
4. Generate files based on the model.

The problem is that while models returned from steps 1 and 3 are cacheable by themselves, combining it with a CompilationProvider makes all that assembly scanning to be performed on every keystroke. 
Which is obviously not great.

As a workaround, I've added a custom [EqualityComparer](https://github.com/Dreamescaper/ServiceScan.SourceGenerator/blob/main/ServiceScan.SourceGenerator/CombinedProviderComparer.cs) for the combined provider, which ignores the compilation and only compares the model. 
It works exactly as I'd expect - the generated file might be stale, since it is regenerated only when the attribute is changed, but it won't detect when mathing types are changed. But it's perfectly fine for my use-case, since it will be regenerated on build anyway.

What are the downsides of such approach (apart from the stale generated files)? Is there any better suggestion here?

**Comment by pekspro** [comment]
Don't combine with CompilationProvider 😄. I've done the same mistake...

From https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md:

Consider the following (incorrect) combine where the basic inputs are combined, then used to generate some source:

```csharp
public void Initialize(IncrementalGeneratorInitializationContext context)
{
    var compilation = context.CompilationProvider;
    var texts = context.AdditionalTextsProvider;

    // Don't do this!
    var combined = texts.Combine(compilation);

    context.RegisterSourceOutput(combined, static (spc, pair) =>
    {
        var assemblyName = pair.Right.AssemblyName;
        // produce source ...
    });
```

Any time the compilation changes, which it will frequently as the user is typing
in the IDE, then `RegisterSourceOutput` will get re-run. Instead, look up the
compilation dependant information first, then combine _that_ with the additional
files:

```csharp
public void Initialize(IncrementalGeneratorInitializationContext context)
{
    var assemblyName = context.CompilationProvider.Select(static (c, _) => c.AssemblyName);
    var texts = context.AdditionalTextsProvider;

    var combined = texts.Combine(assemblyName);

    context.RegisterSourceOutput(combined, (spc, pair) =>
    {
        var assemblyName = pair.Right;
        // produce source ...
    });
}
```

**Reply by Dreamescaper** [reply]
@pekspro 
As I mentioned, I'm adding a custom EqualityProvider specifically to ignore Compilation updates.

Unfortunately, I am not sure how to get the data from Compilation _first_, as I'd need the info about _all the types_.
